### ISSUE TITLE
Feature request: Prune (or skip on creation) empty snapshots

### ISSUE THREAD

**Issue by Debilski** [zrepl/zrepl#728]
We are regularly snapshotting hundreds of people’s home folders and other data, eventually giving us between 10000 (still fine) and 100000 (things are getting much slower) snapshots on one of our storage systems.

While on its own this is not much of a problem, eventually `zfs list` will take a long time to complete (and I presume some of zrepls features that need a global refresh, will also take longer) with that many snapshots. (And some badly written shell-autocompletion scripts suddenly need 2 minutes to fill in a letter.)

In our case a lot of these file systems do not change regularly (especially not during nighttime hours) so we could probably get rid of > 2/3 of these snapshots by simply pruning empty snapshots that do not contain any new information.

Would it be feasible to have something like this built into zrepl? Either as an additional pruning rule or already at snapshot-creation time (just don’t create a new snapshot when nothing has changed).

**Comment by Debilski** [comment]
Ah, great. Didn’t think of using hooks. That sounds like a reasonable solution for now.

**Comment by problame** [comment]
@Halfwalker putting the threshold logic into a hook is a cool idea.

Just for completeness, I assume you use `err_is_fatal=true` in [the zrepl configuration](https://zrepl.github.io/configuration/snapshotting.html#pre-and-post-snapshot-hooks) for you hook?

---

Looking at the OpenZFS source code: the `written` property is cheap to get, the accounting for it is done pro-actively by the kernel anyways.

We could definitely get it as part of the `zfs list` that we already do inside the snapshotter anyways.

The only question is how to hook it up in the config.

I have some longer-arching plans for how to evolve snapshot management, but, in the meantime, maybe we can either
1. extend the hooks framework to allow pre-snapshot-hooks to indicate that the snapshot should be skipped, or
2. introduce the notion of a skip decider into the common snapshotter code.

(1) is pretty much self-describing. It seems quite general, but, I'd prefer capturing common use cases through YAML configuration instead of shell hooks.

Hence, an outline of (2):

The skip decider would run for each filesystem that passes the `filtesystems` filter.
There can be multiple skip deciders that are composable.
* leaf:
  * `command` skip decider to implement arbitrary logic
  * `written` threshold (make your use case built-in)
  * `snapshot_age` threshold (make a snapshot if the last one still present is older than X)
* composition:
  * `and`: logical AND of two skip filters
  * `or`: logical OR of two skip filters

Let me know what you think!

**Comment by Halfwalker** [comment]
Yup, I currently use `err_is_fatal=true` in the zrepl job config for the hook.

I do like the idea of formalizing things in the yaml, but also with the capability to shell out to handle arbitrary conditions.  From your description above it looks like one would define several `leaf` configs, and then the decider chain would list them in order of operation, potentially linking some of them together with logical operators.

That sounds cool, flexible.

You could use the same mechanism possibly for the push jobs.  I can think of several possibilities -
* Only push if a specific interface is available (ie. don't push over laptop wifi, only hardwire from a dock)
* Only push if target is on local network
* Only push between certain hours
* Check conditions on push target (loadavg, network util etc.)

**Comment by reefland** [comment]
When you use the `err_is_fatal=true` zrepl considers it an error condition and generates log noise that can hides real issues.  I opened #715 asking for a way to have hooks report that a snapshot was not taken while also not being an error.

If you think 2/3 of your snapshots can be skipped by using this, that's a lot errors that will flood your log monitoring.
